app/services/foreman/password_hash.rb

```diff
@@ -40,18 +59,25 @@ def hash_secret(password, salt)
       end
     end
 
-    def initialize(implementation = :bcrypt)
-      if implementation == :bcrypt
+    def initialize(implementation = Setting[:password_hash])
+      case implementation&.to_sym
+      when :pbkdf2sha1
+        @implementation = PBKDF2Implementation.new
+      when :bcrypt
         @implementation = BCryptImplementation.new
-      elsif implementation == :sha1
+      when :sha1
         @implementation = SHA1Implementation.new
+      when nil
+        raise(Foreman::Exception.new(N_("Password hash setting was not yet set")))
       else
         raise(Foreman::Exception.new(N_("Unknown password hash method: %s"), implementation))
       end
     end
 
     def self.detect_implementation(password_string)
-      password_string.start_with?('$2') ? :bcrypt : :sha1
+      return :pbkdf2sha1 if password_string.start_with?('$pbkdf2sha1')
+      return :bcrypt if password_string.start_with?('$2')
+      :sha1
```

Comment:
Personally I think this is more readable
```suggestion
      if password_string.start_with?('$pbkdf2sha1')
        :pbkdf2sha1 
      elsif password_string.start_with?('$2')
        :bcrypt
      else
        :sha1
      end
```

Reply:
Sure.